core-v1/src/main/scala/lens/VariantLens.scala

```diff
@@ -147,6 +147,15 @@ object VariantLens extends VariantLensLowPriorityImplicits {
         get = lens.get.andThen(tupler.apply),
       )
   }
+
+  implicit final class AsSumTypeBuilder[A, B](private val lens: VariantLens[A, B]) extends AnyVal {
+
+    /**
+      * Attempt to cast the value of super-type [[B]] as sub-type [[V]] (if possible), otherwise None.
+      */
+    def asCase[V <: B : Typeable]: VariantLens[A, Option[V]] =
+      lens.copy(get = lens.get.andThen(Typeable[V].cast))
```

Comment:
I think I should handle multiple cases at once, otherwise, you would be required to use `when` / `then` / `else` and unify the results. A `Expr.Match` might be a better option... it could also potentially replace the `When` operator as a more generally useful operator, but each branch has a lens that is applied before any optional condition guard.